### DISCUSSION TITLE
Lag in fixing labels after inference across entire video

### DISCUSSION THREAD

**Discussion by gfolw** [talmolab/sleap#1269] [Help!] [answered]
SLEAP worked smoothly and perfectly up until predicting on entire videos. I have large videos that are about 160,000 frames each and am having an issue:

The software lags and it takes about 15 seconds for it to respond to fixing a label. It will take hours for me to fix all the incorrect labels, and for this project, it's important that every label is correct. Is this common, or is it because of my video size?

**Comment by roomrys** [comment, marked as answer]
Hi @gfolw,

Unfortunately, the slowdown scales with the number of labeled frames (including predictions). For this reason, we recommend only predicting/correcting on a subset of frames (in the Labeling Suggestions) until the predictions look ready for production. 

You may still need to make some small corrections after repeating the process above a few times and finally predicting on all frames in the video (which is what it sounds like you doing now?). To make this final correction less cumbersome, I would suggest breaking apart the project on a per video basis. To do this you would need to separate each video into it's own respective project, predict on each video/project using the final model, then make final adjustments to the predictions. Not ideal, but the legwork here will payoff in less lag per correction. 
> Note that you should not separate the project until you are ready for final predictions/corrections as subsequent model training will only include the labeled frames in that specific project.

Let me know if that clears things up.

Thanks,
Liezl

**Reply by jmdelahanty** [reply]
Out of curiosity, is this due to just difficulties in seeking video data and pose points simultaneously? I'm just wondering if there's a fundamental difficulty in software when scrolling through big multi-modal datasets (ie lazily loading lots of data and interfacing with it quickly).

**Reply by roomrys** [reply]
Hi @jmdelahanty,

We try to keep all of our commands from the user in one place: the `CommandContext`. Within that `CommandContext`, we organize the commands based on what they do. The commands we are interested in for the lag are the commands that do any sort of edit. To add an edit command, you need to specify a list of topics that the edit acts upon. We then call the `MainWindow.on_data_update` method after any edit commands are called. The major slowdowns happen when either project_instances or suggestions are edited:
https://github.com/talmolab/sleap/blob/b5aab0edfbcfddce82e8ee02d9bd383604b9de93/sleap/gui/app.py#L1173-L1189

The problem is that we store data in a different way from how we would like to access that data (i.e. looping through a list to count is bound to yield the right result, but is not the fastest). 
> Side Tangent: why not operate our software based on Markov Decision Process instead of running sometimes arbitrary and expensive computations when the effect on the sub-state [i.e. number of instances in frame/project] is fully known for certain commands [i.e. `AddInstance`]

To fix this, we would need to restructure our storage to be more easily accessible (in the ways we know users want to access data). I am really excited to work on optimizing this in the (near 🤞🏼) future, but it'll definitely take some "investment time" to plan a good solution.

Thanks,
Liezl

**Reply by jmdelahanty** [reply]
Wow thanks for such an in depth explanation Liezl! This is really cool to learn about.

**Comment by laurelrr** [comment]
I was having exactly the same issue, with an extremely long lag time (around 15 sec) before the movie frame would update.  I found that I had added some videos that were not compressed.  
Usually, we use a command like: `ffmpeg -i input.asf -vcodec libx264 -crf 23 -preset veryfast -pix_fmt yuv420p output.mp4` before loading the videos into SLEAP.  But I skipped this step with input videos that were already in mp4 format.  
When I went back and ran that conversion, SLEAP's update rate returned to normal.  Is it possible that you ran your predictions on the uncompressed videos?
